power_grid_model_c/power_grid_model/include/power_grid_model/optimizer/tap_position_optimizer.hpp

```diff
@@ -665,8 +672,13 @@ class TapPositionOptimizerImpl<std::tuple<TransformerTypes...>, StateCalculator,
                 if (tap_changed) {
                     break;
                 }
+                iterations_per_rank[++rank_index] = 0;
             }
             if (tap_changed) {
+                if (++iterations_per_rank[rank_index] > 2 * regulator_order[rank_index][0].transformer.tap_range()) {
+                    throw MaxIterationReached{
```

Comment:
i believe that if the optimizer throws, that the cache is not restored. this is a bug in the current implementation because it will already run into this edge case when a `SparseMatrixError` or `IterationDiverge` happens in the solver and it is not able to recover by re-running with the fallback method. Will create a Jira ticket